File: appl/example4/example4.cc
#include <algorithm>
#include <iostream>
#include <stdexcept>
#include <random>
#include <vector>
#include <cmath>
#include <chrono>

// utility functions
#include <frackit/common/math.hh>

// basic geometry types
#include <frackit/geometry/box.hh>
#include <frackit/geometry/sphere.hh>
#include <frackit/geometry/polygon.hh>
#include <frackit/geometry/cylinder.hh>
#include <frackit/geometry/hollowcylinder.hh>
#include <frackit/geometryutilities/getboundingbox.hh>

// headers containing functions to compute lengths/areas/volumes/distances
#include <frackit/distance/distance.hh>
#include <frackit/magnitude/magnitude.hh>

// sampler for points and polygons
#include <frackit/sampling/makeuniformpointsampler.hh>
#include <frackit/sampling/polygonsampler.hh>
#include <frackit/sampling/status.hh>

// constraints to be enforced on the network (distance, angles, etc.)
#include <frackit/entitynetwork/constraints.hh>

// brep utilities of OpenCascade, for boolean operations on shapes
#include <TopoDS_Face.hxx>
#include <frackit/occ/breputilities.hh>

// intersection detection for internal geometry types
#include <frackit/intersection/intersect.hh>
#include <frackit/intersection/emptyintersection.hh>

// builder class for creating networks of entities confined in (sub-)domains
#include <frackit/entitynetwork/networkbuilder.hh>

// writes an entity network to a meshable Gmsh .geo file format
#include <frackit/io/gmshwriter.hh>

int main(int argc, char** argv)
{
    // get start time to compute CPU time later
    const auto start = std::chrono::steady_clock::now();

    //! print welcome message
    std::cout << "\n\n"
              << "#########################################################################\n"
              << "## Example 4: Creation of polygonal entities around a tunnel structure ##\n"
              << "#########################################################################\n"
              << "\n\n\n";

    // Define some types used here
    using namespace Frackit;
    using ctype = double;
    using Box = Frackit::Box<ctype>;
    using Sphere = Frackit::Sphere<ctype>;
    using Cylinder = Frackit::Cylinder<ctype>;
    using HollowCylinder = Frackit::HollowCylinder<ctype>;
    using Polygon = Frackit::Polygon<ctype, 3>;

    using Point = typename Polygon::Point;
    using Circle = typename Cylinder::Circle;
    using Direction = typename Cylinder::Direction;
    using Vector = typename Direction::Vector;

    /////////////////////////////
    // 1. Construct the domain //
    /////////////////////////////

    // In this example, we consider a fraction of a geological layer
    // around a tunnel structure. To this end, we construct the geological
    // layer as box, and intersect it with a sphere in order to restrict
    // the model domain to a smaller region.

    // the layer has a thickness of 40m, and initially we consider 100m lateral extent
    const Box layer(-50, -50, -20.0, 50, 50, 20.0);
    const Sphere domainSphere(Point(0.0, 0.0, 0.0), /*radius*/25.0);

    // a tunnel with a radius of 5m crosses the domain
    const ctype tunnelRadius = 5.0;
    const Direction tunnelDirection(Vector(1.0, 0.0, 0.0));
    const Circle tunnelBase(Point(-25.0, 0.0, 0.0), tunnelDirection, tunnelRadius);
    const Cylinder tunnel(tunnelBase, /*height*/50.0);

    // the domain is the layer constrained to the domainSphere and cut by the tunnel
    const auto layerSphereCut = OCCUtilities::intersect(OCCUtilities::getShape(layer),
                                                        OCCUtilities::getShape(domainSphere),
                                                        /*eps*/1e-6);
    auto solids = OCCUtilities::getSolids(layerSphereCut);
    if (solids.size() != 1) throw std::runtime_error("Intersection operation with domainSphere should yield a single solid");

    const auto tunnelCut = OCCUtilities::cut(solids[0], OCCUtilities::getShape(tunnel), /*eps*/1e-6);
    solids = OCCUtilities::getSolids(tunnelCut);
    if (solids.size() != 1) throw std::runtime_error("Cut operation with tunnel should yield a single solid");

    // get the domain shape
    const auto& domain = solids[0];

    // get the OCC reprensentation of the sphere surface
    const auto sphereFaces = OCCUtilities::getFaces(OCCUtilities::getShape(domainSphere));
    if (sphereFaces.size() != 1) throw std::runtime_error("We expect the sphere to consist of a single face");
    const auto& sphereSurfaceShape = sphereFaces[0];

    // get the OCC representation of tunnel surface
    const auto tunnelSurfaceShape = OCCUtilities::getShape(tunnel.lateralFace());


    ////////////////////////////////////////////////////////////////////////
    // 2. Make sampler classes to randomly sample points (entity centers) //
    //    and polygons as entities (using the sampled center points)      //
    ////////////////////////////////////////////////////////////////////////

    // we use the default sampler types, thus, default distributions (see traits classes)
    // This means, normal distributions for angles and uniform distributions for the size
    using PolygonSampler = Frackit::PolygonSampler<3, ctype>;

    // lambda function to construct polygon sampler with given point sampler & orientation
    auto getSampler = [] (const auto& pointSampler, unsigned int orientation) -> PolygonSampler
    {
        using NormalDistro = std::normal_distribution<ctype>;
        auto sizeDistro = std::uniform_real_distribution<ctype>(3.0, 6.0);
        auto numCornerDistro = std::uniform_int_distribution<int>(4, 9);

        switch (orientation)
        {
            case 1:
                return PolygonSampler(pointSampler,                                   // sampler for polygon centers
                                      NormalDistro(toRadians(15.0), toRadians(10.0)), // strike angle: mean value & standard deviation
                                      NormalDistro(toRadians(90.0), toRadians(10.0)), // dip angle: mean value & standard deviation
                                      /*strikeLength/*/sizeDistro, /*dipLength*/sizeDistro, numCornerDistro);
            case 2:
                return PolygonSampler(pointSampler,                                  // sampler for polygon centers
                                      NormalDistro(toRadians(0.0), toRadians(10.0)), // strike angle: mean value & standard deviation
                                      NormalDistro(toRadians(0.0), toRadians(10.0)), // dip angle: mean value & standard deviation
                                      /*strikeLength/*/sizeDistro, /*dipLength*/sizeDistro, numCornerDistro);
            case 3:
                return PolygonSampler(pointSampler,                                    // sampler for polygon centers
                                      NormalDistro(toRadians(-75.0), toRadians(10.0)), // strike angle: mean value & standard deviation
                                      NormalDistro(toRadians(90.0), toRadians(10.0)),  // dip angle: mean value & standard deviation
                                      /*strikeLength/*/sizeDistro, /*dipLength*/sizeDistro, numCornerDistro);
            default:
                throw std::runtime_error("Unsupported orientation index");
        }
    };


    ///////////////////////////////////////////////////////////////////////
    // 3. Define constraints that should be fulfilled among the entities //
    //    of different orientations.                                     //
    ///////////////////////////////////////////////////////////////////////

    // Define constraints between entities of orientation 1
    using Constraints = EntityNetworkConstraints<ctype>;
    Constraints constraints1;
    constraints1.setMinDistance(1.0);
    constraints1.setMinIntersectingAngle(toRadians(25.0));
    constraints1.setMinIntersectionMagnitude(0.1);
    constraints1.setMinIntersectionDistance(0.1);

    // Define constraints between entities of orientation 2
    // we want to enforce larger spacing between those entities
    auto constraints2 = constraints1;
    constraints2.setMinDistance(0.5);

    // Define constraints between entities of orientation 3
    auto constraints3 = constraints1;
    constraints2.setMinDistance(0.15);

    // Define constraints between entities of different sets
    auto constraintsOnOther = constraints1;
    constraintsOnOther.setMinDistance(0.25);
    constraintsOnOther.setMinIntersectingAngle(toRadians(30.0));

    // Define constraints to the domain boundary
    auto constraintsOnBoundary = constraints1;
    constraintsOnBoundary.setMinIntersectingAngle(toRadians(10.0));

    // When polygons intersect, the intersection might be very close to
    // one of the corner points. We want to avoid small length scales caused
    // by this, and this lambda provides the check for it.
    auto producesSmallLengthScale = [] (const auto& geometry, const auto& is) -> bool
    {
        // lambda to check if the intersection is too close (< 1mm) to a vertex of the geometry
        auto isTooClose = [&is] (const auto& v) { return computeDistance(v, is) < 1e-3; };

        const auto shape = OCCUtilities::getShape(geometry);
        const auto vertices = OCCUtilities::getVertices(shape);
        return std::any_of(vertices.begin(), vertices.end(), isTooClose);
    };


    ///////////////////////////
    // 4. Network generation //
    ///////////////////////////
    const std::size_t numTargetEntities = 75;
    std::vector<TopoDS_Face> entitiesSet1; entitiesSet1.reserve(numTargetEntities);
    std::vector<TopoDS_Face> entitiesSet2; entitiesSet2.reserve(numTargetEntities);
    std::vector<TopoDS_Face> entitiesSet3; entitiesSet3.reserve(numTargetEntities);

    // Define ids for the two entity sets
    const Id idSet1(1);
    const Id idSet2(2);
    const Id idSet3(3);

    // In a first step, we want to populate entities of orientation 1 around
    // the tunnel, and require each entity to be intersecting with it. To this end,
    // we sample center points within a hollow cylinder around the tunnel.
    SamplingStatus status; status.setTargetCount(idSet1, numTargetEntities);
    const HollowCylinder sampleCylinder1(tunnel.bottomFace().center(),
                                         tunnel.direction(),
                                         tunnel.radius(),
                                         tunnel.radius() + 2.0,
                                         tunnel.height());

    ctype area = 0.0;
    auto polygonSampler1 = getSampler(makeUniformPointSampler(sampleCylinder1), /*orientationIdx*/1);
    while (!status.finished())
    {
        const auto candidate = polygonSampler1();
        const auto is = intersect(tunnel.lateralFace(), candidate);

        // we only want entities that intersect the tunnel
        if (isEmptyIntersection(is))
        { status.increaseRejectedCounter(); continue; }

        // avoid small length scales produced by the intersection
        if (producesSmallLengthScale(candidate, is))
        { status.increaseRejectedCounter(); continue; }

        // compute the part contained in the domain
        const auto containedShape = OCCUtilities::intersect(OCCUtilities::getShape(candidate), domain, /*eps*/1e-6);
        const auto containedFaces = OCCUtilities::getFaces(containedShape);

        // skip everything that does not lead to a single contained face
        if (containedFaces.size() != 1)
        { status.increaseRejectedCounter(); continue; }

        // if too little of the candidate is inside the domain, reject it
        auto containedFace = containedFaces[0];
        const auto containedFaceArea = computeMagnitude(containedFace);
        if (containedFaceArea < 0.25*candidate.area())
        { status.increaseRejectedCounter(); continue; }

        // check constraints to other entities and domain boundary
        if (!constraints1.evaluate(entitiesSet1, containedFace))
        { status.increaseRejectedCounter(); continue; }
        if (!constraintsOnBoundary.evaluate(tunnelSurfaceShape, containedFace))
        { status.increaseRejectedCounter(); continue; }
        if (!constraintsOnBoundary.evaluate(sphereSurfaceShape, containedFace))
        { status.increaseRejectedCounter(); continue; }

        area += containedFaceArea;
        entitiesSet1.emplace_back(std::move(containedFace));
        status.increaseCounter(idSet1);
        status.print();
    }

    std::cout << "\n\n\n"
              << "## Step 2: for each entity so far, try to generate an intersecting entity with orientation 2\n"
              << "\n\n";

    // after rejecting 500 candidates for an entity, we move to the next one
    const std::size_t maxTriesForEntity = 500;
    auto printSkipMessage = [&maxTriesForEntity] ()
    {
        std::cout << "\t\t -- Skipped search for this entity after "
                  << maxTriesForEntity << " unsuccessful tries. --" << std::endl;
    };

    status.reset();
    status.setTargetCount(idSet2, entitiesSet1.size());
    for (const auto& e : entitiesSet1)
    {
        using std::sqrt;

        // sample an entity of orientation 2 in the vicinity of e
        const auto rawBBox = getBoundingBox(e);
        const auto charLength = sqrt(computeMagnitude(e));
        const Box sampleBox(rawBBox.xMin()-charLength, rawBBox.yMin()-charLength, rawBBox.zMin()-charLength,
                            rawBBox.xMax()+charLength, rawBBox.yMax()+charLength, rawBBox.zMax()+charLength);

        bool accepted = false;
        std::size_t tryCount = 0;
        auto sampler = getSampler(makeUniformPointSampler(sampleBox), /*orientationIdx*/2);

        while (!accepted && tryCount <= maxTriesForEntity)
        {
            tryCount++;
            const auto candidate = sampler();
            const auto is = intersect(candidate, e);

            // we only want intersecting entities
            if (isEmptyIntersection(is))
            { status.increaseRejectedCounter(); continue; }

            // avoid small length scales produced by the intersection
            if (producesSmallLengthScale(candidate, is))
            { status.increaseRejectedCounter(); continue; }
            if (producesSmallLengthScale(e, is))
            { status.increaseRejectedCounter(); continue; }

            // compute the part contained in the domain
            const auto containedShape = OCCUtilities::intersect(OCCUtilities::getShape(candidate), domain, /*eps*/1e-6);
            const auto containedFaces = OCCUtilities::getFaces(containedShape);

            // skip everything that does not lead to a single contained face
            if (containedFaces.size() != 1)
            { status.increaseRejectedCounter(); continue; }

            // if too little of the candidate is inside the domain, reject it
            auto containedFace = containedFaces[0];
            const auto containedFaceArea = computeMagnitude(containedFace);
            if (containedFaceArea < 0.25*candidate.area())
            { status.increaseRejectedCounter(); continue; }

            // check constraints to other entities and domain boundary
            if (!constraintsOnOther.evaluate(entitiesSet1, containedFace))
            { status.increaseRejectedCounter(); continue; }
            if (!constraints2.evaluate(entitiesSet2, containedFace))
            { status.increaseRejectedCounter(); continue; }
            if (!constraintsOnBoundary.evaluate(sphereSurfaceShape, containedFace))
            { status.increaseRejectedCounter(); continue; }
            if (!constraintsOnBoundary.evaluate(tunnelSurfaceShape, containedFace))
            { status.increaseRejectedCounter(); continue; }

            accepted = true;
            area += containedFaceArea;
            entitiesSet2.emplace_back(std::move(containedFace));
            status.increaseCounter(idSet2);
            status.print();
        }

        if (tryCount >= maxTriesForEntity) printSkipMessage();
    }

    std::cout << "\n\n\n"
              << "## Step 3: for each entity of orientation 2, try to generate an intersecting entity with orientation 3\n"
              << "\n\n";

    status.reset();
    status.setTargetCount(idSet3, entitiesSet2.size());
    for (const auto& e : entitiesSet2)
    {
        using std::sqrt;

        // sample an entity of orientation 2 in the vicinity of e
        const auto rawBBox = getBoundingBox(e);
        const auto charLength = sqrt(computeMagnitude(e));
        const Box sampleBox(rawBBox.xMin()-charLength, rawBBox.yMin()-charLength, rawBBox.zMin()-charLength,
                            rawBBox.xMax()+charLength, rawBBox.yMax()+charLength, rawBBox.zMax()+charLength);

        auto sampler = getSampler(makeUniformPointSampler(sampleBox), /*orientationIdx*/3);
        bool accepted = false;
        std::size_t tryCount = 0;

        while (!accepted && tryCount <= maxTriesForEntity)
        {
            tryCount++;
            auto candidate = sampler();
            const auto is = intersect(candidate, e);

            // we only want intersecting entities
            if (isEmptyIntersection(is))
            { status.increaseRejectedCounter(); continue; }

            // avoid small length scales produced by the intersection
            if (producesSmallLengthScale(candidate, is))
            { status.increaseRejectedCounter(); continue; }
            if (producesSmallLengthScale(e, is))
            { status.increaseRejectedCounter(); continue; }

            // compute the part contained in the domain
            const auto containedShape = OCCUtilities::intersect(OCCUtilities::getShape(candidate), domain, /*eps*/1e-6);
            const auto containedFaces = OCCUtilities::getFaces(containedShape);

            // skip everything that does not lead to a single contained face
            if (containedFaces.size() != 1)
            { status.increaseRejectedCounter(); continue; }

            // if too little of the candidate is inside the domain, reject it
            auto containedFace = containedFaces[0];
            const auto containedFaceArea = computeMagnitude(containedFace);
            if (containedFaceArea < 0.25*candidate.area())
            { status.increaseRejectedCounter(); continue; }

            // check constraints to other entities and domain boundary
            if (!constraintsOnOther.evaluate(entitiesSet1, containedFace))
            { status.increaseRejectedCounter(); continue; }
            if (!constraintsOnOther.evaluate(entitiesSet2, containedFace))
            { status.increaseRejectedCounter(); continue; }
            if (!constraints3.evaluate(entitiesSet3, containedFace))
            { status.increaseRejectedCounter(); continue; }
            if (!constraintsOnBoundary.evaluate(sphereSurfaceShape, containedFace))
            { status.increaseRejectedCounter(); continue; }
            if (!constraintsOnBoundary.evaluate(tunnelSurfaceShape, containedFace))
            { status.increaseRejectedCounter(); continue; }

            accepted = true;
            area += containedFaceArea;
            entitiesSet3.emplace_back(std::move(containedFace));
            status.increaseCounter(idSet3);
            status.print();
        }

        if (tryCount >= maxTriesForEntity) printSkipMessage();
    }

    // print the final entity density
    const auto domainVolume = computeMagnitude(domain);
    const auto numEntities = entitiesSet1.size()+entitiesSet2.size()+entitiesSet3.size();
    std::cout << "\nCreated a network consisting of " << numEntities << " entities\n";
    std::cout << "Volume of the domain: " << domainVolume << std::endl;
    std::cout << "Area of the network: " << area << " m²"
              << ", which corresponds to a density of " << area/domainVolume << " m²/m³" << std::endl;


    /////////////////////////////////////////////////////////////////////////////////////////
    // 5. The entities of the network have been created. We can now construct the network. //
    /////////////////////////////////////////////////////////////////////////////////////////

    // construct and write a contained network, i.e. write out both network and domain.
    std::cout << "Building and writing network" << std::endl;
    ContainedEntityNetworkBuilder<ctype> builder;

    // add sub-domains
    builder.addConfiningSubDomain(domain,     Id(1));
    builder.addSubDomainEntities(entitiesSet1, Id(1));
    builder.addSubDomainEntities(entitiesSet2, Id(1));
    builder.addSubDomainEntities(entitiesSet3, Id(1));

    // now we can build and write out the network in Gmsh file format
    GmshWriter gmshWriter(builder.build());
    gmshWriter.write("network", /*sizeAtEntities*/0.5, /*sizeInDomain*/5.0);

    // print time it took to generate the network
    const auto end = std::chrono::steady_clock::now();
    const auto duration = std::chrono::duration<double>(end-start).count();
    std::cout << "Overall CPU time was " << duration << " seconds." << std::endl;

    return 0;
}

File: appl/example4/example4.py
import sys
from time import process_time
startTime = process_time()

# print welcome message
print("\n\n"
      "#########################################################################\n"
      "## Example 4: Creation of polygonal entities around a tunnel structure ##\n"
      "#########################################################################\n"
      "\n\n")

###########################
# 1.Construct the domain ##
###########################

# In this example, we consider a fraction of a geological layer
# around a tunnel structure. To this end, we construct the geological
# layer as box, and intersect it with a sphere in order to restrict
# the model domain to a smaller region.
from frackit.geometry import Box, Sphere, Cylinder, Direction, Vector, Circle, Point

# the layer has a thickness of 40m, and initially we consider 100m lateral extent
layer = Box(-50.0, -50.0, -20.0, 50.0, 50.0, 20.0)
domainSphere = Sphere(Point(0.0, 0.0, 0.0), 25.0)

# a tunnel with a radius of 5m crosses the domain
tunnelRadius = 5.0
tunnelDirection = Direction(Vector(1.0, 0.0, 0.0))
tunnelBase = Circle(Point(-25.0, 0.0, 0.0), tunnelDirection, tunnelRadius)
tunnel = Cylinder(tunnelBase, 50.0)

# the domain is the layer constrained to the domainSphere and cut by the tunnel
from frackit.occutilities import getShape, getSolids, getFaces, cut
from frackit.occutilities import intersect as intersectShapes
solids = getSolids( intersectShapes(getShape(layer), getShape(domainSphere), 1e-6) )
if len(solids) != 1: sys.exit("Intersection operation with domainSphere should yield a single solid")

solids = getSolids( cut(solids[0], getShape(tunnel), 1e-6) )
if len(solids) != 1: sys.exit("Cut operation with tunnel should yield a single solid")

domain = solids[0]
tunnelSurfaceShape = getShape(tunnel.lateralFace())
sphereSurfaceShape = getFaces(getShape(domainSphere))
if len(sphereSurfaceShape) != 1: sys.exit("Sphere is expected to contain a single face")
sphereSurfaceShape = sphereSurfaceShape[0]

########################################################################
## 2. Make sampler classes to randomly sample points (entity centers) ##
##    and polygons as entities (using the sampled center points)      ##
########################################################################

# creates a uniform sampler within an interval
import random
def uniformSampler(a, b):
    def sample(): return random.uniform(a, b)
    return sample

# creates a sampler from a normal distribution with the given mean and std deviation
def normalSampler(mean, stdDev):
    def sample(): return random.gauss(mean, stdDev)
    return sample

# creates a sampler from an integer uniform distribution
def choiceSampler(a, b):
    def sample(): return random.choice([v for v in range(a, b+1)])
    return sample

# function to create polygon samplers for the different orientations
from frackit.common import toRadians
from frackit.sampling import PolygonSampler

sizeDistro = uniformSampler(3.0, 6.0)
numCornersDistro = choiceSampler(4, 9)

def getSampler(pointSampler, orientation):
    if orientation == 1:
        return PolygonSampler(pointSampler,                                    # sampler for polygon centers
                              normalSampler(toRadians(15.0), toRadians(10.0)), # strike angle: mean value & standard deviation
                              normalSampler(toRadians(90.0), toRadians(10.0)), # dip angle: mean value & standard deviation
                              sizeDistro, sizeDistro, numCornersDistro)        # strike & dip length, number of corners
    elif orientation == 2:
        return PolygonSampler(pointSampler,                                   # sampler for polygon centers
                              normalSampler(toRadians(0.0), toRadians(10.0)), # strike angle: mean value & standard deviation
                              normalSampler(toRadians(0.0), toRadians(10.0)), # dip angle: mean value & standard deviation
                              sizeDistro, sizeDistro, numCornersDistro)       # strike & dip length, number of corners
    elif orientation == 3:
        return PolygonSampler(pointSampler,                                     # sampler for polygon centers
                              normalSampler(toRadians(-75.0), toRadians(10.0)), # strike angle: mean value & standard deviation
                              normalSampler(toRadians(90.0), toRadians(10.0)),  # dip angle: mean value & standard deviation
                              sizeDistro, sizeDistro, numCornersDistro)         # strike & dip length, number of corners
    else:
        import sys
        sys.exit("Unsupported orientation index")


#######################################################################
## 3. Define constraints that should be fulfilled among the entities ##
##    of different orientations.                                     ##
#######################################################################

from frackit.entitynetwork import EntityNetworkConstraints, EntityNetworkConstraintsMatrix

# constructs default constraints for this test to be modified after
def makeDefaultConstraints():
    c = EntityNetworkConstraints()
    c.setMinDistance(1.0)
    c.setMinIntersectingAngle(toRadians(25.0))
    c.setMinIntersectionMagnitude(0.1)
    c.setMinIntersectionDistance(0.1)
    return c

# Define constraints between entities of orientation 1
constraints1 = makeDefaultConstraints()

# Define constraints between entities of orientation 2
# we want to enforce larger spacing between those entities
constraints2 = makeDefaultConstraints()
constraints2.setMinDistance(0.5)

# Define constraints between entities of orientation 3
constraints3 = makeDefaultConstraints()
constraints3.setMinDistance(0.15)

# Define constraints between entities of different sets
constraintsOnOther = makeDefaultConstraints()
constraintsOnOther.setMinDistance(0.25)
constraintsOnOther.setMinIntersectingAngle(toRadians(30.0))

# Moreover, we define constraints w.r.t. the domain boundary
constraintsOnBoundary = makeDefaultConstraints()
constraintsOnBoundary.setMinIntersectingAngle(toRadians(10.0))

# When polygons intersect, the intersection might be very close to
# one of the corner points. We want to avoid small length scales caused
# by this, and this lambda provides the check for it.
def producesSmallLengthScale(geometry, isection):

    from frackit.geometry import computeDistance
    from frackit.occutilities import getVertices
    verts = getVertices(getShape(geometry))
    if any((computeDistance(v, isection) < 1e-3 for v in verts)): return True
    return False

###########################
## 4. Network generation ##
###########################

# dictionary to store entities of the three orientations in
from frackit.common import Id
setId1 = Id(1); setId2 = Id(2); setId3 = Id(3);
entitySets = {setId1: [], setId2: [], setId3: []}

# In a first step, we want to populate entities of orientation 1 around
# the tunnel, and require each entity to be intersecting with it. To this end,
# we sample center points within a hollow cylinder around the tunnel.
from frackit.sampling import SamplingStatus
from frackit.geometry import HollowCylinder
numTargetEntities = 75
status = SamplingStatus()
status.setTargetCount(setId1, numTargetEntities)
sampleCylinder1 = HollowCylinder(tunnel.bottomFace().center(), tunnel.direction(),
                                 tunnel.radius(), tunnel.radius() + 2.0, tunnel.height())

from frackit.geometry import computeMagnitude, getBoundingBox
from frackit.geometry import intersect
from frackit.occutilities import getFaces
from frackit.sampling import makeUniformPointSampler
print('## Step 1: generation of entities of orientation 1 around the tunnel\n\n\n')

area = 0.0;
polygonSampler1 = getSampler(makeUniformPointSampler(sampleCylinder1), 1)
while not status.finished():
    candidate = polygonSampler1()
    isection = intersect(tunnel.lateralFace(), candidate)

    # skip entities that don't intersect the tunnel
    if all( "Empty" in g.name() for g in isection ):
        status.increaseRejectedCounter(); continue;

    # avoid small length scales produced by the intersections
    if producesSmallLengthScale(candidate, isection):
         status.increaseRejectedCounter(); continue;

    # compute the part contained in the domain
    containedShape = intersectShapes(getShape(candidate), domain, 1e-6)
    containedFaces = getFaces(containedShape)

    # skip everything that does not lead to a single contained face
    if len(containedFaces) != 1: status.increaseRejectedCounter(); continue;

    # if too little of the candidate is inside the domain, reject it
    containedFace = containedFaces[0];
    containedFaceArea = computeMagnitude(containedFace);
    if containedFaceArea < 0.25*candidate.area():
        status.increaseRejectedCounter(); continue;

    # check constraints to other entities and domain boundary
    if not constraints1.evaluate(entitySets[setId1], containedFace):
        status.increaseRejectedCounter(); continue;
    if not constraintsOnBoundary.evaluate(tunnelSurfaceShape, containedFace):
        status.increaseRejectedCounter(); continue;
    if not constraintsOnBoundary.evaluate(sphereSurfaceShape, containedFace):
        status.increaseRejectedCounter(); continue;

    area += containedFaceArea
    entitySets[setId1].append(containedFace)
    status.increaseCounter(setId1)
    status.print()

print('\n\n\n## Step 2: for each entity so far, try to generate an intersecting entity with orientation 2\n\n\n')

# after rejecting 500 candidates for an entity, we move to the next one
maxTriesForEntity = 500;
def printSkipMessage ():
    print('\t\t -- Skipped search for this entity after ' + \
          str(maxTriesForEntity) + ' unsuccessful tries. --\n')

from math import sqrt
status.reset()
status.setTargetCount(setId2, len(entitySets[setId1]))
for e in entitySets[setId1]:

    # sample an entity of orientation 2 in the vicinity of e
    rawBBox = getBoundingBox(e);
    charLength = sqrt(computeMagnitude(e));
    sampleBox = Box(rawBBox.xMin()-charLength, rawBBox.yMin()-charLength, rawBBox.zMin()-charLength,
                    rawBBox.xMax()+charLength, rawBBox.yMax()+charLength, rawBBox.zMax()+charLength);

    tryCount = 0;
    accepted = False;
    sampler = getSampler(makeUniformPointSampler(sampleBox), 2);

    while not accepted and tryCount <= maxTriesForEntity:
        tryCount += 1
        candidate = sampler()
        isection = intersect(candidate, e)

        # we only want intersecting entities
        if all( "Empty" in g.name() for g in isection ):
            status.increaseRejectedCounter(); continue;

        # avoid small length scales produced by the intersections
        if producesSmallLengthScale(candidate, isection):
             status.increaseRejectedCounter(); continue;
        if producesSmallLengthScale(e, isection):
             status.increaseRejectedCounter(); continue;

        # compute the part contained in the domain
        containedShape = intersectShapes(getShape(candidate), domain, 1e-6)
        containedFaces = getFaces(containedShape)

        # skip everything that does not lead to a single contained face
        if len(containedFaces) != 1: status.increaseRejectedCounter(); continue;

        # if too little of the candidate is inside the domain, reject it
        containedFace = containedFaces[0];
        containedFaceArea = computeMagnitude(containedFace);
        if containedFaceArea < 0.25*candidate.area():
            status.increaseRejectedCounter(); continue;

        # check constraints to other entities and domain boundary
        if not constraintsOnOther.evaluate(entitySets[setId1], containedFace):
            status.increaseRejectedCounter(); continue;
        if not constraints2.evaluate(entitySets[setId2], containedFace):
            status.increaseRejectedCounter(); continue;
        if not constraintsOnBoundary.evaluate(sphereSurfaceShape, containedFace):
            status.increaseRejectedCounter(); continue;
        if not constraintsOnBoundary.evaluate(tunnelSurfaceShape, containedFace):
            status.increaseRejectedCounter(); continue;

        accepted = True;
        area += containedFaceArea
        entitySets[setId2].append(containedFace)
        status.increaseCounter(setId2)
        status.print()

    if tryCount >= maxTriesForEntity: printSkipMessage()


print('\n\n\n## Step 3: for each entity of orientation 2, try to generate an intersecting entity with orientation 3\n\n\n')
status.reset()
status.setTargetCount(setId3, len(entitySets[setId2]))

for e  in entitySets[setId2]:

    # sample an entity of orientation 2 in the vicinity of e
    rawBBox = getBoundingBox(e);
    charLength = sqrt(computeMagnitude(e));
    sampleBox = Box(rawBBox.xMin()-charLength, rawBBox.yMin()-charLength, rawBBox.zMin()-charLength,
                    rawBBox.xMax()+charLength, rawBBox.yMax()+charLength, rawBBox.zMax()+charLength);

    tryCount = 0;
    accepted = False;
    sampler = getSampler(makeUniformPointSampler(sampleBox), 3);

    while not accepted and tryCount <= maxTriesForEntity:
        tryCount += 1
        candidate = sampler()
        isection = intersect(candidate, e)

        # we only want intersecting entities
        if all( "Empty" in g.name() for g in isection ):
            status.increaseRejectedCounter(); continue;

        # avoid small length scales produced by the intersections
        if producesSmallLengthScale(candidate, isection):
             status.increaseRejectedCounter(); continue;
        if producesSmallLengthScale(e, isection):
             status.increaseRejectedCounter(); continue;

        # compute the part contained in the domain
        containedShape = intersectShapes(getShape(candidate), domain, 1e-6)
        containedFaces = getFaces(containedShape)

        # skip everything that does not lead to a single contained face
        if len(containedFaces) != 1: status.increaseRejectedCounter(); continue;

        # if too little of the candidate is inside the domain, reject it
        containedFace = containedFaces[0];
        containedFaceArea = computeMagnitude(containedFace);
        if containedFaceArea < 0.25*candidate.area():
            status.increaseRejectedCounter(); continue;

        # check constraints to other entities and domain boundary
        if not constraintsOnOther.evaluate(entitySets[setId1], containedFace):
            status.increaseRejectedCounter(); continue;
        if not constraintsOnOther.evaluate(entitySets[setId2], containedFace):
            status.increaseRejectedCounter(); continue;
        if not constraints3.evaluate(entitySets[setId3], containedFace):
            status.increaseRejectedCounter(); continue;
        if not constraintsOnBoundary.evaluate(sphereSurfaceShape, containedFace):
            status.increaseRejectedCounter(); continue;
        if not constraintsOnBoundary.evaluate(tunnelSurfaceShape, containedFace):
            status.increaseRejectedCounter(); continue;

        accepted = True;
        area += containedFaceArea
        entitySets[setId3].append(containedFace)
        status.increaseCounter(setId3)
        status.print()

    if tryCount >= maxTriesForEntity: printSkipMessage()

# print the final entity density
domainVolume = computeMagnitude(domain);
numEntities = len(entitySets[setId1]) + len(entitySets[setId2]) + len(entitySets[setId3])
print("\nCreated a network consisting of {:d} entities.".format(numEntities))
print("Volume of the domain: {:.2f}.".format(domainVolume))
print("Area of the network: {:.2f} m², which corresponds to a density of {:.2f} m²/m³".format(area, area/domainVolume))

########################################################################################
## 5. The entities of the network have been created. We can now construct the network ##
########################################################################################

print("Building and writing network")
from frackit.entitynetwork import ContainedEntityNetworkBuilder
builder = ContainedEntityNetworkBuilder();

# add sub-domains
builder.addConfiningSubDomain(domain,    Id(1))
for id in entitySets: builder.addSubDomainEntities(entitySets[id], Id(1))

# now we can build and write out the network in Gmsh file format
from frackit.io import GmshWriter
gmshWriter = GmshWriter(builder.build());
gmshWriter.write("network", 0.5, 5.0)

stopTime = process_time()
print("Overall CPU time was {:.2f} seconds.".format(stopTime-startTime))
